### ISSUE TITLE
C20 Coroutines: sketching a minimal async framework

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10478]
C++20 Coroutines: sketching a minimal async framework<br>
<br>
https://ift.tt/KOzC30w<br>
<br>
Jeremy Ong<br>
<br>

<div><div><div><div>
<div>
<p>C++20 ✨coroutines✨ are here! Having spent some 20-odd hours with it, I am by no means an expert on the feature,
but I wanted to jot down my initial impressions and provide some pointers looking to get their feet wet and
try their hand at implementing a coroutine-powered async framework themselves. Spoiler-alert: for engineers that
need to parallelize their code, in particular, engineers used to needing to “jobify” CPU work using lambdas or
function objects, I think coroutines are a godsend in writing expressive code. I have no doubt that they will
be a staple in future AAA game engines, simulation software, and more. What follows below is a pretty
abbreviated tour of coroutines, as I really don’t think you can truly “grok” it without writing a lot of
code yourself, stepping in with a debugger, littering your code with dozens of <code>printf</code>s, etc. However,
I hope that this post both motivates the use of C++20 coroutines, and also demonstrates that once you
<em>do</em> understand the language facilities for coroutines more, you can do a surprising amount with not a lot
of code.</p>
<h2>Basic Concepts</h2>
<p>This post is going to be overwhelmingly more “applied” in nature, as opposed to delving into the nitty gritty.
That said, we still need to get on the same page about some of the terms you’ll see when working with coroutines
in C++20. Note that to <em>use</em> a coroutine based framework written by someone else, you can actually get away with
surprisingly little knowledge (although there are some gotchas as we will see).</p>
<p>First up is the <em>coroutine</em> itself. A coroutine <em>is a</em> function which can be suspended and resumed (as opposed to
executing from start to finish). As a function, coroutines can be invoked by other coroutines, or by other functions.
As a technicality, the main function itself cannot be a coroutine (you can’t suspend from it). A function becomes
a coroutine the moment you suspend from within its body by invoking <code>co_await</code> or <code>co_yield</code>. Note that if you call
a coroutine <code>c</code> (remember, coroutines are just functions) from a function <code>f</code>, the function <code>f</code> is <em>not</em> a coroutine
unless you <code>co_await</code> the result of <code>c</code>. In other words, calling a coroutine isn’t sufficient to make the caller a
coroutine. Something must happen in the body of the caller which may cause it to suspend for the caller to be a
coroutine.</p>
<p>The next concept to understand is the <code>promise_type</code>. A <code>promise_type</code> is an interface type that can be thought of
as a control object the programmer can use to inject custom behaviour during the lifecycle of a coroutine. Some
sample coroutine lifecycle events are:</p>
<ul>
<li>An allocation is happening to store the state of a coroutine (activation frame plus any additional data)</li>
<li>Coroutine is initially created</li>
<li>Coroutine is about to suspend</li>
<li>Coroutine has just finished suspending</li>
<li>Coroutine is completed and is now returning a value</li>
</ul>
<p>Each of these events corresponds to a method you can implement to specify custom behavior.</p>
<p>The final concept to grok when working with coroutines is the <code>awaitable</code>. An awaitable is anything that can
be the argument of the <code>co_await</code> unary operator. In order to qualify as an awaitable such that <code>co_await a</code>
is valid, either, <code>a</code> has a type with <code>operator co_await</code> implemented, or <code>a</code> has a type that implements the
<code>awaitable</code> trait, or, there is an <code>await_transform</code> method present on the parent coroutine (remember, an await
happens in the body of a coroutine). Regardless of the mechanism, what is consumed by the <code>co_await</code> operator
is a object (produced via any of the means above) which possesses the following methods:</p>
<ul>
<li>An <code>await_ready</code> that describes if the suspend needs to happen at all (useful for short-circuiting awaits
for which the associated work was already completed)</li>
<li>An <code>await_resume</code> which specifies what happens once the await is over</li>
<li>An <code>await_suspend</code> which is passed a <code>coroutine_handle</code> that can be used to resume the parent coroutine later</li>
</ul>
<h2>Motivating example</h2>
<p>Suppose we want to read a file from disk, decode it, and then use the results. Furthermore, suppose we want
to run some work in parallel while the file is being read and while the file is being decoded. Here’s what
this code might look like with coroutines.</p>
<div><div><pre><code><span>task</span><span>&lt;</span><span>std</span><span>::</span><span>string</span><span>&gt;</span> <span>read_file</span><span>(</span><span>std</span><span>::</span><span>filesystem</span><span>::</span><span>path</span> <span>const</span><span>&amp;</span> <span>path</span><span>)</span> <span>{</span>
  <span>// File reading operation</span>
  <span>// Assumes a win32 implementation using overlapped IO, but</span>
  <span>// this same concept applies to epoll-style file reading as well.</span>
  <span>HANDLE</span> <span>completion_event</span> <span>=</span> <span>win_init_read_file</span><span>(</span><span>path</span><span>);</span>

  <span>co_await</span> <span>completion_event</span><span>;</span>
<span>}</span>

<span>task</span><span>&lt;</span><span>std</span><span>::</span><span>vector</span><span>&lt;</span><span>uint8_t</span><span>&gt;&gt;</span> <span>decode_png</span><span>(</span><span>std</span><span>::</span><span>string</span> <span>const</span><span>&amp;</span> <span>data</span><span>)</span> <span>{</span>
  <span>// Indicate that this function should run as a separate task</span>
  <span>// on a different thread</span>
  <span>co_await</span> <span>suspend</span><span>();</span>

  <span>std</span><span>::</span><span>vector</span><span>&lt;</span><span>uint8_t</span><span>&gt;</span> <span>out</span><span>;</span>
  <span>// PNG decoding code</span>
  <span>co_return</span> <span>out</span><span>;</span>
<span>}</span>

<span>task</span><span>&lt;&gt;</span> <span>some_expensive_function1</span><span>()</span>
<span>{</span>
  <span>co_await</span> <span>suspend</span><span>();</span>

  <span>// Some expensive code</span>
<span>}</span>

<span>task</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>some_expensive_function2</span><span>()</span>
<span>{</span>
  <span>co_await</span> <span>suspend</span><span>();</span>

  <span>// Some expensive code that produces an integer result</span>
  <span>co_return</span> <span>result</span><span>;</span>
<span>}</span>

<span>task</span><span>&lt;&gt;</span> <span>some_expensive_function3</span><span>(</span><span>int</span><span>)</span>
<span>{</span>
  <span>co_await</span> <span>suspend</span><span>();</span>

  <span>// Some expensive code</span>
<span>}</span>

<span>void</span> <span>some_loading_function</span><span>(</span><span>std</span><span>::</span><span>filesystem</span><span>::</span><span>path</span> <span>const</span><span>&amp;</span> <span>path</span><span>)</span> <span>{</span>
  <span>task</span><span>&lt;</span><span>std</span><span>::</span><span>string</span><span>&gt;</span> <span>read_task</span> <span>=</span> <span>read_file</span><span>(</span><span>path</span><span>);</span>

  <span>// Do two other useful things in parallel while the file</span>
  <span>// is being read</span>
  <span>task</span><span>&lt;&gt;</span> <span>task1</span> <span>=</span> <span>some_expensive_function1</span><span>();</span>
  <span>// The second task produces an integral result</span>
  <span>task</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>task2</span> <span>=</span> <span>some_expensive_function2</span><span>();</span>
  
  <span>// Await the file read and pass the data to the png decoder</span>
  <span>task</span><span>&lt;</span><span>std</span><span>::</span><span>vector</span><span>&lt;</span><span>uint8_t</span><span>&gt;&gt;</span> <span>png_task</span> <span>=</span>
    <span>decode_png</span><span>(</span><span>co_await</span> <span>read_task</span><span>);</span>

  <span>int</span> <span>task2_result</span> <span>=</span> <span>co_await</span> <span>task2</span><span>;</span>

  <span>// Do other useful thing while png is decoding using the</span>
  <span>// result from task2</span>
  <span>some_expensive_function3</span><span>(</span><span>task2_result</span><span>);</span>

  <span>// Ensure task1 is also finished before proceeding</span>
  <span>co_await</span> <span>task1</span><span>;</span>

  <span>do_something_with_png</span><span>(</span><span>co_await</span> <span>png_task</span><span>);</span>
<span>}</span>

</code></pre></div></div>
<p>When looking at the implementation of <code>some_loading_function</code>, it’s important to imagine how
difficult this would be to implement with a traditional task-graph style API. Most of these
functions would need to be written as function objects, or we’d have lambdas littered throughout.
However the task graph is implemented, the resulting code would read necessarily <em>non-linearly</em>,
especially in this example where we are essentially forking a function and joining it in a non-trivial
manner. Suffice it to say, I believe the interface shown in the example above will likely be terser
and more readable. So, let’s implement it!</p>
<p>There are <em>three</em> elements that we need to implement to make the above code snippet a reality.</p>
<ol>
<li>First, we need some way to await for some OS event to signal completion. That is, we need to
wait for something to complete outside of our program.</li>
<li>Second, we need a way to <em>suspend execution</em> of a function in the current thread in such a way
that this function is resumed automatically in a different thread. This is how we will schedule
multiple tasks to complete in parallel.</li>
<li>Finally, we need a way to await for work that we’ve enqueued via the suspension routine above.</li>
</ol>
<p>The underlying “substrate” on which our coroutines will run also needs to be defined. Let’s call this
the “Central Task Processor” or CTP for short. It’s interface should be as follows:</p>
<div><div><pre><code><span>class</span> <span>CTP</span> <span>{</span>

<span>public:</span>
  <span>// For simplicity, we'll assume a global singleton instance is available</span>
  <span>static</span> <span>CTP</span><span>&amp;</span> <span>instance</span><span>();</span>

  <span>// The constructor should spawn thread_count threads and begin the run loop of each thread</span>
  <span>CTP</span><span>(</span><span>int</span> <span>thread_count</span><span>);</span>

  <span>// When a coroutine is enqueued, the coroutine is passed to one of the CTP threads, and</span>
  <span>// eventually coroutine.resume() should be invoked when the thread is ready. You can</span>
  <span>// implement work stealing here to "steal" coroutines on busy threads from idle ones, or</span>
  <span>// any other load balancing/scheduling scheme of your choice.</span>
  <span>//</span>
  <span>// NOTE: This should (obviously) be threadsafe</span>
  <span>void</span> <span>enqueue</span><span>(</span><span>std</span><span>::</span><span>coroutine_handle</span><span>&lt;&gt;</span> <span>coroutine</span><span>);</span>
<span>};</span>
</code></pre></div></div>
<p>This simple interface is all we need to indicate that a coroutine that was previously suspended is
now ready to resume. The <code>CTP</code> will presumably find a thread ready to accept work and resume the coroutine
there, or enqueue it for later. Now, let’s handle the actual coroutine suspension code.</p>
<h2>Awaiting an OS event</h2>
<p>In order to <code>co_await</code> for some OS handle that will be completed later, we need to implement <code>operator co_await</code>,
taking as its sole parameter, the argument to the unary <code>co_await</code> operator. Thankfully, this
is really quite simple. Again, using windows as an example, we might implement this as follows:</p>
<div><div><pre><code><span>struct</span> <span>event_awaiter</span> <span>{</span>
  <span>HANDLE</span> <span>event</span><span>;</span>
  <span>bool</span> <span>await_ready</span><span>()</span> <span>const</span> <span>noexcept</span> <span>{</span>
    <span>// Check if we need to bother suspending at all by seeing if the</span>
    <span>// event was already signaled with a non-blocking wait.</span>
    <span>return</span> <span>WaitForSingleObject</span><span>(</span><span>event</span><span>,</span> <span>0</span><span>)</span> <span>==</span> <span>WAIT_OBJECT_0</span><span>;</span>
  <span>}</span>

  <span>void</span> <span>await_suspend</span><span>(</span><span>std</span><span>::</span><span>coroutine_handle</span><span>&lt;&gt;</span> <span>coroutine</span><span>)</span> <span>const</span> <span>noexcept</span> <span>{</span>
    <span>// The coroutine handle passed here can be copied elsewhere and resumed</span>
    <span>// when the event is signaled. Here, we spawn a dedicated thread for</span>
    <span>// demonstration purposes, but you should have a dedicated low-priority</span>
    <span>// thread to queue waiting events to.</span>
    <span>std</span><span>::</span><span>thread</span> <span>thread</span><span>{[</span><span>coroutine</span><span>,</span> <span>event</span><span>]()</span> <span>noexcept</span> <span>{</span>
      <span>WaitForSingleObject</span><span>(</span><span>event</span><span>,</span> <span>INFINITE</span><span>);</span>

      <span>// The CTP will call coroutine.resume() on an available thread now</span>
      <span>CTP</span><span>::</span><span>instance</span><span>().</span><span>enqueue</span><span>(</span><span>coroutine</span><span>);</span>
    <span>}};</span>
    <span>thread</span><span>.</span><span>detach</span><span>();</span>
  <span>}</span>

  <span>void</span> <span>await_resume</span><span>()</span> <span>noexcept</span> <span>{</span>
    <span>// This is called after the coroutine is resumed in the async thread</span>
    <span>printf</span><span>(</span><span>"Event signaled, resuming on thread %i</span><span>\n</span><span>"</span><span>,</span> <span>get_thread_id</span><span>());</span>
  <span>}</span>
<span>};</span>
</code></pre></div></div>
<p>To make our event <code>HANDLE</code> a valid argument to the <code>co_await</code> operator, we also need to add
an overload for returning the <code>event_awaiter</code> defined above.</p>
<div><div><pre><code><span>inline</span> <span>event_awaiter</span> <span>operator</span> <span>co_await</span><span>(</span><span>HANDLE</span> <span>event</span><span>)</span> <span>noexcept</span>
<span>{</span>
  <span>return</span> <span>{</span><span>event</span><span>};</span>
<span>}</span>
</code></pre></div></div>
<p>Now, we can do something like the following:</p>
<div><div><pre><code><span>// Retrieve an event that has yet to be signaled</span>
<span>HANDLE</span> <span>event</span> <span>=</span> <span>some_routine</span><span>();</span>

<span>// Assuming the event is unsignaled, this task will now suspend (await_suspend</span>
<span>// will be called above). The thread that this code originally ran on is now</span>
<span>// free to resume a different coroutine enqueued earlier (if any) or simply</span>
<span>// wait until a task is ready.</span>
<span>co_await</span> <span>event</span><span>;</span>

<span>// At this point, await_resume() above has been called and we are now running</span>
<span>// on a potentially different CTP thread.</span>
<span>some_other_routine</span><span>();</span>
</code></pre></div></div>
<p>Neat!</p>
<h2>Suspend code execution</h2>
<p>So, the code above lets us suspend a coroutine and resume it when the OS tells
us it’s ready. Let’s now figure out how to suspend a task to resume immediately
on a different CTP thread. Again, we’ll do this using an awaitable.</p>
<div><div><pre><code><span>inline</span> <span>auto</span> <span>suspend</span><span>()</span> <span>noexcept</span> <span>{</span>
  <span>struct</span> <span>awaiter</span> <span>{</span>
    <span>// Unlike the OS event case, there's no case where we suspend and the work</span>
    <span>// is immediately ready</span>
    <span>bool</span> <span>await_ready</span><span>()</span> <span>const</span> <span>noexcept</span> <span>{</span> <span>return</span> <span>false</span><span>;</span> <span>}</span>

    <span>// Since await_ready() always returns false, when suspend is called, we will</span>
    <span>// always immediately suspend and call this function (which enqueues the coroutine</span>
    <span>// for immediate reactivation on a different thread)</span>
    <span>void</span> <span>await_suspend</span><span>(</span><span>std</span><span>::</span><span>coroutine_handle</span><span>&lt;&gt;</span> <span>coroutine</span><span>)</span> <span>noexcept</span> <span>{</span>
      <span>CTP</span><span>::</span><span>instance</span><span>().</span><span>enqueue</span><span>(</span><span>coroutine</span><span>);</span>
    <span>}</span>

    <span>void</span> <span>await_resume</span><span>()</span> <span>const</span> <span>noexcept</span> <span>{</span>
      <span>printf</span><span>(</span><span>"Suspended task now running on thread $i</span><span>\n</span><span>"</span><span>,</span> <span>get_thread_id</span><span>());</span>
    <span>}</span>
  <span>};</span>
  <span>return</span> <span>awaiter</span><span>{};</span>
<span>}</span>
</code></pre></div></div>
<p>This is even more straightforward than the OS event case. In action, this looks like the
following:</p>
<div><div><pre><code><span>// Note, we do not yet know what a "task" is. That will be introduced next</span>
<span>task</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>async_routine</span><span>()</span> <span>{</span>
  <span>// Suspend this current coroutine and enqueue it to our CTP</span>
  <span>co_await</span> <span>suspend</span><span>();</span>

  <span>// At this point, we're running on a thread that resumed this coroutine's handle</span>
  <span>// In this case, we compute some integer and "co_return" it</span>
  <span>co_return</span> <span>some_expensive_thing</span><span>();</span>
<span>}</span>
</code></pre></div></div>
<h2>Introducing the “task”</h2>
<p>So far, we’ve implemented two “awaiters” which describe what happens when the coroutine
itself is suspended and resumed. However, this only addresses one side of the story.
Namely, we have not described anywhere in code what should happen to the thread <em>which began
the execution of the suspended coroutine</em>. Take the <code>async_routine</code> function we wrote
above which produces an <code>int</code> result asynchronously. How should we define where the result
is written to? Also, how does the caller of <code>async_routine</code> wait for it to finish? The
answer is with the <code>promise_type</code> concept which we will wrap in an abstraction called a <code>task</code>.</p>
<div><div><pre><code><span>template</span> <span>&lt;</span><span>typename</span> <span>T</span><span>&gt;</span>
<span>struct</span> <span>task</span> <span>{</span>
  <span>// The return type of a coroutine must contain a nested struct or type alias called `promise_type`</span>
  <span>struct</span> <span>promise_type</span> <span>{</span>
    <span>// Keep a coroutine handle referring to the parent coroutine if any. That is, if we</span>
    <span>// co_await a coroutine within another coroutine, this handle will be used to continue</span>
    <span>// working from where we left off.</span>
    <span>std</span><span>::</span><span>coroutine_handle</span><span>&lt;&gt;</span> <span>precursor</span><span>;</span>

    <span>// Place to hold the results produced by the coroutine</span>
    <span>T</span> <span>data</span><span>;</span>

    <span>// Invoked when we first enter a coroutine. We initialize the precursor handle</span>
    <span>// with a resume point from where the task is ultimately suspended</span>
    <span>task</span> <span>get_return_object</span><span>()</span> <span>noexcept</span>
    <span>{</span>
      <span>return</span> <span>{</span><span>std</span><span>::</span><span>coroutine_handle</span><span>&lt;</span><span>promise_type</span><span>&gt;::</span><span>from_promise</span><span>(</span><span>*</span><span>this</span><span>)};</span>
    <span>}</span>

    <span>// When the caller enters the coroutine, we have the option to suspend immediately.</span>
    <span>// Let's choose not to do that here</span>
    <span>std</span><span>::</span><span>suspend_never</span> <span>initial_suspend</span><span>()</span> <span>const</span> <span>noexcept</span> <span>{</span> <span>return</span> <span>{};</span> <span>}</span>

    <span>// If an exception was thrown in the coroutine body, we would handle it here</span>
    <span>void</span> <span>unhandled_exception</span><span>()</span> <span>{}</span>

    <span>// The coroutine is about to complete (via co_return or reaching the end of the coroutine body).</span>
    <span>// The awaiter returned here defines what happens next</span>
    <span>auto</span> <span>final_suspend</span><span>()</span> <span>const</span> <span>noexcept</span> <span>{</span>
      <span>struct</span> <span>awaiter</span> <span>{</span>
        <span>// Return false here to return control to the thread's event loop. Remember that we're</span>
        <span>// running on some async thread at this point.</span>
        <span>bool</span> <span>await_ready</span><span>()</span> <span>const</span> <span>noexcept</span> <span>{</span> <span>return</span> <span>false</span><span>;</span> <span>}</span>

        <span>void</span> <span>await_resume</span><span>()</span> <span>const</span> <span>noexcept</span> <span>{}</span>

        <span>// Returning a coroutine handle here resumes the coroutine it refers to (needed for</span>
        <span>// continuation handling). If we wanted, we could instead enqueue that coroutine handle</span>
        <span>// instead of immediately resuming it by enqueuing it and returning void.</span>
        <span>std</span><span>::</span><span>coroutine_handle</span><span>&lt;&gt;</span> <span>await_suspend</span><span>(</span><span>std</span><span>::</span><span>coroutine_handle</span><span>&lt;</span><span>promise_type</span><span>&gt;</span> <span>h</span><span>)</span> <span>{</span>
          <span>auto</span> <span>precursor</span> <span>=</span> <span>h</span><span>.</span><span>promise</span><span>().</span><span>precursor</span><span>;</span>
          <span>if</span> <span>(</span><span>precursor</span><span>)</span>
          <span>{</span>
            <span>return</span> <span>precursor</span><span>;</span>
          <span>}</span>
          <span>return</span> <span>std</span><span>::</span><span>noop_coroutine</span><span>();</span>
        <span>}</span>
      <span>};</span>
      <span>return</span> <span>awaiter</span><span>{};</span>

      <span>// When the coroutine co_returns a value, this method is used to publish the result</span>
      <span>void</span> <span>return_value</span><span>(</span><span>T</span> <span>value</span><span>)</span> <span>noexcept</span>
      <span>{</span>
        <span>data</span> <span>=</span> <span>std</span><span>::</span><span>move</span><span>(</span><span>value</span><span>);</span>
      <span>}</span>
    <span>}</span>
  <span>};</span>

  <span>// The following methods make our task type conform to the awaitable concept, so we can</span>
  <span>// co_await for a task to complete</span>

  <span>bool</span> <span>await_ready</span><span>()</span> <span>const</span> <span>noexcept</span> <span>{</span>
    <span>// No need to suspend if this task has no outstanding work</span>
    <span>return</span> <span>handle</span><span>.</span><span>done</span><span>();</span>
  <span>}</span>

  <span>T</span> <span>await_resume</span><span>()</span> <span>const</span> <span>noexcept</span> <span>{</span>
    <span>// The returned value here is what `co_await our_task` evaluates to</span>
    <span>return</span> <span>std</span><span>::</span><span>move</span><span>(</span><span>handle</span><span>.</span><span>promise</span><span>().</span><span>data</span><span>);</span>
  <span>}</span>

  <span>void</span> <span>await_suspend</span><span>(</span><span>std</span><span>::</span><span>coroutine_handle</span><span>&lt;&gt;</span> <span>coroutine</span><span>)</span> <span>const</span> <span>noexcept</span> <span>{</span>
    <span>// The coroutine itself is being suspended (async work can beget other async work)</span>
    <span>// Record the argument as the continuation point when this is resumed later. See</span>
    <span>// the final_suspend awaiter on the promise_type above for where this gets used</span>
    <span>handle</span><span>.</span><span>promise</span><span>().</span><span>precursor</span> <span>=</span> <span>coroutine</span><span>;</span>
  <span>}</span>

  <span>// This handle is assigned to when the coroutine itself is suspended (see await_suspend above)</span>
  <span>std</span><span>::</span><span>coroutine_handle</span><span>&lt;</span><span>promise_type</span><span>&gt;</span> <span>handle</span><span>;</span>
<span>};</span>
</code></pre></div></div>
<p>OK, there’s understandably a lot to unpack here, so let’s look at a simple function
and establish a chronology for what happens and why.</p>
<div><div><pre><code><span>task</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>async_routine</span><span>()</span> <span>{</span>
  <span>co_await</span> <span>suspend</span><span>();</span>

  <span>co_return</span> <span>some_expensive_thing</span><span>();</span>
<span>}</span>

<span>task</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>another_async_routine</span><span>()</span> <span>{</span>
  <span>// Enqueue 3 invocations of async_routine such that they are all running</span>
  <span>// in parallel, assuming we have 3 threads available</span>
  <span>// The current thread enters `async_routine`, then encounters the co_await of the</span>
  <span>// suspend() above. This yields control back so we can spawn the next async_routine.</span>
  <span>// In each case, get_return_object is called to construct a task with memory to</span>
  <span>// store the returned fresult. Furthermore, the coroutine handle to each of these</span>
  <span>// invocations is stored so that when async_routine suspends internally, this</span>
  <span>// caller can continue where it left off.</span>
  <span>task</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>t1</span> <span>=</span> <span>async_routine</span><span>();</span>
  <span>task</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>t2</span> <span>=</span> <span>async_routine</span><span>();</span>
  <span>task</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>t3</span> <span>=</span> <span>async_routine</span><span>();</span>

  <span>// Because t2 is itself an awaitable, a few things happen here. First, we check</span>
  <span>// in await_ready above if the coroutine handle associated with t2 has already</span>
  <span>// finalized. If so, we don't need to suspend and can just retrieve the result</span>
  <span>// directly (internally invokes `return_value`). If t2 hasn't finished yet, we</span>
  <span>// assign the precursor handle to the current resume point and suspend. When t2</span>
  <span>// finishes later, its final_suspend will find this precursor and resume here.</span>
  <span>// Note that it's possible that after this line, we end up running on a different</span>
  <span>// thread (the thread that finished t2)</span>
  <span>int</span> <span>r2</span> <span>=</span> <span>co_await</span> <span>t2</span><span>;</span>

  <span>do_something_else</span><span>();</span>

  <span>// Resume points work in the middle of expressions as well</span>
  <span>co_return</span> <span>((</span><span>co_await</span> <span>t1</span><span>)</span> <span>+</span> <span>r2</span> <span>+</span> <span>(</span><span>co_await</span> <span>t3</span><span>));</span>
<span>}</span>
</code></pre></div></div>
<p>It’s very difficult to describe in exact terms how each “control” method for all
the different awaitables involved coordinate to enable the code above to work. In
general terms though, this is roughly what’s happening:</p>
<ol>
<li>Every time we enter a coroutine (denoted as a function which returns a <code>task</code>),
the caller constructs the <code>task</code> object which contains storage space for data to be
returned by the coroutine (you’ll need to roll a <code>task&lt;void&gt;</code> specialization as well
if you don’t want that)</li>
<li>The task also stores a <code>precursor</code> which refers to the <em>resume point</em>. Initially,
the <code>precursor</code> is just the point in code directly after the <code>task</code> construction.</li>
<li>The task itself can suspend, either by awaiting another task, or by awaiting one of
the two awaitables we made before.</li>
<li>When a task completes (by working it’s way to the end of its function body or by
encountering a <code>co_return</code> statement), it immediately resumes the <code>precursor</code>, which
presumably suspended because of this task.</li>
</ol>
<h2>How do I await a coroutine from something that isn’t a coroutine?</h2>
<p>This is a fair question. At some point, the entry into a coroutine must come from
something that <em>isn’t</em> a coroutine. After all, the standard dictates that <code>main</code> must
not be a coroutine. To do this, you’ll need to implement your own <code>await</code> method which
waits on a <code>condition_variable</code> or OS event directly. You can store this <code>condition_variable</code>
on the <code>task</code> type, then set it in the task’s <code>final_suspend</code> method for example.</p>
<p>In my codebase, this is referred to as a <code>jtask</code> (short for “joinable task”) and it is
otherwise identical to the task above except for this additional signaling behavior.</p>
<h2>Some implementation details I glossed over</h2>
<ul>
<li>You can override <code>operator new</code> and <code>operator delete</code> to manage allocations of your task
types yourself. Note that I don’t believe these are guaranteed to be called if the compiler
can elide the allocation altogether, so don’t implement these overrides with visible side-effects</li>
<li>When suspending a coroutine, remember to ensure that pointers and references used in the
body of the coroutine must have a valid lifetime extended until the coroutine exits. This
is “obvious” but potentially is a pitfall (similar to capturing pointers and references in a lambda)</li>
<li>If doing this implementation yourself, you can also apply a similar pattern in how we handled
continuations and suspensions here to handle <code>co_yield</code> expressions.</li>
<li>You will likely need a <code>task&lt;void&gt;</code> specialization in order to handle coroutines that execute
but don’t return anything</li>
<li>My own code has additional state in the <code>task</code> type to handle things like tracking history
and file and line information (simplifies debugging). Also, the code is instrumented to support
labels in profiling tools like Pix.</li>
</ul>
<h2>Conclusion</h2>
<p>C++20 coroutines are undoutedly powerful. I didn’t go into all the compiler optimizations
that can kick in to do fancy things like elide heap allocations, but there are significant
advantages to migrating to use coroutines instead of the typical task-graph abstraction.
However, as a powerful abstraction (with extreme amounts of flexibility), C++20 coroutines
are undeniably difficult to learn, at least at first. To get to my current level of proficiency,
I essentially made dozens of example projects with copious amounts of <code>printf</code> statements
to help me understand the control flow. I highly recommend doing this exercise yourself, as
you may find you would prefer to make different design choices than those offered here.</p>
<p>As I mentioned before, this post isn’t meant to be exhaustive <em>at all</em>, but should hopefully
serve to whet your appetite in terms of what is possible and introduce the general terminology.
My engine (which I can’t publish for reasons at the moment) benefits a great deal from coroutines
with a task and thread pool implementation measuring only several hundreds of lines long!
The real benefit however is in the calling code. After writing this once, I’ve rarely needed
to think much about the mechanics. The only caution needed perhaps is to remember that the
thread a coroutine starts on is not necessarily the same thread that it ends on (as implemented here).
Hopefully, you’ll derive some benefits from C++20 coroutines in the near future also!</p>
</div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Jeremy's Blog <br>
<br>
December 26, 2023 at 10:58AM
